datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/counter.py

```diff
@@ -0,0 +1,516 @@
+        if not isinstance(instance_counts, dict):
+            raise ConfigTypeError(f'Option `instance_counts` for performance object `{self.name}` must be a mapping')
+
+        unknown_count_types = set(instance_counts) - {'total', 'monitored', 'unique'}
+        if unknown_count_types:
+            raise ConfigValueError(
+                f'Option `instance_counts` for performance object `{self.name}` has unknown types: '
+                f'{", ".join(sorted(unknown_count_types))}'
+            )
+
+        for count_type, metric_name in instance_counts.items():
+            if not isinstance(metric_name, str):
+                raise ConfigTypeError(
+                    f'Metric name for count type `{count_type}` of option `instance_counts` for performance object '
+                    f'`{self.name}` must be a string'
+                )
+
+        self.instance_count_total_metric = instance_counts.get('total')
+        self.instance_count_monitored_metric = instance_counts.get('monitored')
+        self.instance_count_unique_metric = instance_counts.get('unique')
+
+        self.counters_config = config.get('counters', [])
+        if not isinstance(self.counters_config, list):
+            raise ConfigTypeError(f'Option `counters` for performance object `{self.name}` must be an array')
+        elif not self.counters_config:
+            raise ConfigValueError(f'Option `counters` for performance object `{self.name}` is required')
+
+        # We'll configure on the first run because it's necessary to figure out whether the
+        # object contains single or multi-instance counters, see:
+        # https://docs.microsoft.com/en-us/windows/win32/perfctrs/object-and-counter-design
+        self.counters = []
+        self.has_multiple_instances = False
+
+    def collect(self):
+        for counter in self.counters:
+            counter.collect()
+
+    def refresh(self):
+        # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhenumobjectitemsa
+        # http://timgolden.me.uk/pywin32-docs/win32pdh__EnumObjectItems_meth.html
+        counters, instances = win32pdh.EnumObjectItems(
+            None, self.connection.server, self.name, win32pdh.PERF_DETAIL_WIZARD
+        )
+        if not self.counters:
+            self._configure_counters(counters, instances)
+
+        counters = set(counters)
+        total_instance_count = 0
+        monitored_instance_count = 0
+        unique_instance_counts = defaultdict(int)
+        for instance in instances:
+            total_instance_count += 1
+            if self._instance_excluded(instance):
+                continue
+
+            monitored_instance_count += 1
+            unique_instance_counts[instance] += 1
+
+        if self.has_multiple_instances:
+            if self.instance_count_total_metric is not None:
+                self.check.gauge(self.instance_count_total_metric, total_instance_count, tags=self.tags)
```

Comment:
should these be gauges or monotonic_count to reflect the total count?

Reply:
gauges b/c number of instances can go up or down